File: src/differ_api/services/diff.py
from __future__ import annotations

from differ_api.repositories.comments import CommentRepository
from differ_api.repositories.repos import RepoRepository
from differ_api.repositories.sessions import SessionRepository
from differ_api.services.errors import BadRequestError, NotFoundError
from differ_api.services.validators import validate_path_filters


class DiffService:
    def __init__(
        self,
        *,
        repo_repository: RepoRepository,
        session_repository: SessionRepository,
        comment_repository: CommentRepository,
        git_client,
    ) -> None:
        self.repo_repository = repo_repository
        self.session_repository = session_repository
        self.comment_repository = comment_repository
        self.git_client = git_client

    def compare(
        self,
        *,
        session_id: str,
        base_ref: str | None,
        head_ref: str | None,
        base_commit: str | None,
        head_commit: str | None,
        path_filters: list[str] | None,
    ) -> dict:
        session = self.session_repository.get(session_id)
        if session is None:
            raise NotFoundError("Session not found")
        repo = self.repo_repository.get(session.repo_id)
        if repo is None:
            raise NotFoundError("Repository not found")

        effective_base_ref = base_ref or session.base_ref
        effective_head_ref = head_ref or session.head_ref
        effective_filters = validate_path_filters(path_filters if path_filters is not None else session.path_filters)

        if not effective_base_ref and not base_commit:
            raise BadRequestError("A base ref or base commit is required")
        if not effective_head_ref and not head_commit:
            raise BadRequestError("A head ref or head commit is required")

        if effective_base_ref:
            self.git_client.ensure_ref_exists(repo.path, effective_base_ref)
        if effective_head_ref:
            self.git_client.ensure_ref_exists(repo.path, effective_head_ref)

        resolved_base_commit = self.git_client.resolve_revision(
            repo.path,
            ref=effective_base_ref,
            commit=base_commit,
        )
        resolved_head_commit = self.git_client.resolve_revision(
            repo.path,
            ref=effective_head_ref,
            commit=head_commit,
        )

        diff = self.git_client.diff(
            repo.path,
            base_commit=resolved_base_commit,
            head_commit=resolved_head_commit,
            path_filters=effective_filters,
        )
        comments = self.comment_repository.list(
            session_id=session_id,
            head_commit_sha=resolved_head_commit,
        )
        return {
            "base": self.git_client.get_commit_summary(repo.path, resolved_base_commit, effective_base_ref),
            "head": self.git_client.get_commit_summary(repo.path, resolved_head_commit, effective_head_ref),
            "stats": diff["stats"],
            "files": diff["files"],
            "comments": comments,
        }

File: electron/main/services/diff.ts
import { RepoRepository } from '../repositories/repo'
import { SessionRepository } from '../repositories/session'
import { CommentRepository } from '../repositories/comment'
import { GitClient } from '../git/client'
import { NotFoundError, BadRequestError } from './errors'
import { validatePathFilters } from './validators'
import type { CommentRow } from '../db/types'

export interface DiffResponse {
  base: {
    ref: string | null
    commit: string
    author_name: string
    author_email: string
    authored_at: string
    subject: string
  }
  head: {
    ref: string | null
    commit: string
    author_name: string
    author_email: string
    authored_at: string
    subject: string
  }
  stats: {
    files_changed: number
    additions: number
    deletions: number
  }
  files: import('../git/adapter').DiffFile[]
  comments: CommentRow[]
}

export class DiffService {
  constructor(
    private repoRepo: RepoRepository,
    private sessionRepo: SessionRepository,
    private commentRepo: CommentRepository,
    private gitClient: GitClient
  ) {}

  async compare(data: {
    session_id: string
    base_ref?: string | null
    head_ref?: string | null
    base_commit?: string | null
    head_commit?: string | null
    path_filters?: string[] | null
  }): Promise<DiffResponse> {
    const session = this.sessionRepo.getById(data.session_id)
    if (!session) throw new NotFoundError('Session not found')

    const repo = this.repoRepo.getById(session.repo_id)
    if (!repo) throw new NotFoundError('Repository not found')

    const effectiveBaseRef = data.base_ref ?? session.base_ref
    const effectiveHeadRef = data.head_ref ?? session.head_ref
    const effectiveFilters = validatePathFilters(
      data.path_filters !== undefined && data.path_filters !== null
        ? data.path_filters
        : JSON.parse(session.path_filters_json)
    )

    if (!effectiveBaseRef && !data.base_commit) {
      throw new BadRequestError('A base ref or base commit is required')
    }
    if (!effectiveHeadRef && !data.head_commit) {
      throw new BadRequestError('A head ref or head commit is required')
    }

    if (effectiveBaseRef) {
      await this.gitClient.ensureRefExists(repo.path, effectiveBaseRef)
    }
    if (effectiveHeadRef) {
      await this.gitClient.ensureRefExists(repo.path, effectiveHeadRef)
    }

    const resolvedBaseCommit = await this.gitClient.resolveRevision(repo.path, {
      ref: effectiveBaseRef,
      commit: data.base_commit ?? null,
    })
    const resolvedHeadCommit = await this.gitClient.resolveRevision(repo.path, {
      ref: effectiveHeadRef,
      commit: data.head_commit ?? null,
    })

    const diff = await this.gitClient.diff(repo.path, {
      baseCommit: resolvedBaseCommit,
      headCommit: resolvedHeadCommit,
      pathFilters: effectiveFilters,
    })

    const comments = this.commentRepo.list({
      session_id: data.session_id,
      head_commit_sha: resolvedHeadCommit,
    })

    const [baseSummary, headSummary] = await Promise.all([
      this.gitClient.getCommitSummary(repo.path, resolvedBaseCommit, effectiveBaseRef ?? null),
      this.gitClient.getCommitSummary(repo.path, resolvedHeadCommit, effectiveHeadRef ?? null),
    ])

    return {
      base: baseSummary,
      head: headSummary,
      stats: diff.stats,
      files: diff.files,
      comments,
    }
  }
}
